maggy/core/experimentdriver.py

```diff
@@ -344,38 +471,67 @@ def _update_result(self, trial):
-            self.result = {'best_id': trial_id, 'best_val': metric,
-                'best_hp': param_string, 'worst_id': trial_id,
-                'worst_val': metric, 'worst_hp': param_string,
-                'avg': metric, 'metric_list': [metric], 'num_trials': 1,
-                'early_stopped': 0}
-
-            if trial.early_stop:
-                self.result['early_stopped'] += 1
-
-            return
-        if self.direction == 'max':
+        if self.experiment_type == 'optimization':
+            # First finalized trial
+            if self.result.get('best_id', None) is None:
+                self.result = {
+                    'best_id': trial_id, 'best_val': metric,
+                    'best_hp': param_string, 'worst_id': trial_id,
+                    'worst_val': metric, 'worst_hp': param_string,
+                    'avg': metric, 'metric_list': [metric], 'num_trials': 1,
+                    'early_stopped': 0,
+                }
+
+                if trial.early_stop:
+                    self.result['early_stopped'] += 1
+
+                return
+            if self.direction == 'max':
+                if metric > self.result['best_val']:
+                    self.result['best_val'] = metric
+                    self.result['best_id'] = trial_id
+                    self.result['best_hp'] = param_string
+                if metric < self.result['worst_val']:
+                    self.result['worst_val'] = metric
+                    self.result['worst_id'] = trial_id
+                    self.result['worst_hp'] = param_string
+            elif self.direction == 'min':
+                if metric < self.result['best_val']:
+                    self.result['best_val'] = metric
+                    self.result['best_id'] = trial_id
+                    self.result['best_hp'] = param_string
+                if metric > self.result['worst_val']:
+                    self.result['worst_val'] = metric
+                    self.result['worst_id'] = trial_id
+                    self.result['worst_hp'] = param_string
+
+        elif self.experiment_type == 'ablation':
+
+            # pop function values and trial_type from parameters, since we don't need them
+            param_string.pop('dataset_function', None)
+            param_string.pop('model_function', None)
+            # First finalized trial
+            if self.result.get('best_id', None) is None:
+                self.result = {
+                    'best_id': trial_id, 'best_val': metric, 'best_config': param_string,
+                    'worst_id': trial_id, 'worst_val': metric, 'worst_config': param_string,
+                    'avg': metric, 'metric_list': [metric], 'num_trials': 1,
+                    'early_stopped': 0,
+                    # TODO earlystop included temporarily for compatibility with sparkmagic, remove before release
+                }
+                return
+
+            # for ablation we always consider 'max' as "the direction"
```

Comment:
For now let's keep it that way, since eventually this is only used for showing the trial with `best_val` in the notebook, and the assumption is that the users are interested in accuracy. But the point totally makes sense, and they might want to use e.g. loss so `best_val` should correspond to the lowest loss.

Anyhow, I will explain this in the documentation and will add this as a TODO to the board, so we can come up with a better way of handling this.